File: library_zap/src/main/java/com/dolin/zap/format/DateFileFormatter.java
package com.dolin.zap.format;

import android.text.TextUtils;

import com.dolin.zap.entity.Level;
import com.dolin.zap.internal.IFormatter;
import com.dolin.zap.util.LevelUtils;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

/**
 * @author #Suyghur.
 * Created on 2021/05/19
 */
public class DateFileFormatter implements IFormatter {

    private SimpleDateFormat simpleDateFormat;
    private Date date = null;
    private String lastDateFormatted = "";
    private StringBuffer stringBuffer;
    private int timeLength = 0;

    public DateFileFormatter() {
        this("yyyy-MM-dd HH:mm:ss");
    }

    public DateFileFormatter(String pattern) {
        simpleDateFormat = new SimpleDateFormat(pattern, Locale.getDefault());
        stringBuffer = new StringBuffer();
        //重置秒数
        Calendar.getInstance().set(Calendar.SECOND, 0);
        date = Calendar.getInstance().getTime();
    }

    private void resetTimePrefix() {
        if (stringBuffer == null) {
            return;
        }
        if (stringBuffer.length() > 0) {
            stringBuffer.delete(0, stringBuffer.length());
        }
        timeLength = stringBuffer.append(lastDateFormatted).append(" ").length();
    }

    private String formatString(Level level, String tag, String msg) {
        if (stringBuffer == null) {
            return "";
        }
        if (stringBuffer.length() > timeLength) {
            stringBuffer.delete(timeLength, stringBuffer.length());
        }
        return stringBuffer.append("[").append(LevelUtils.getShortLevelName(level)).append("]")
                .append(tag).append(": ").append(msg).append("\n").toString();
    }

    @Override
    public synchronized String format(Level level, String tag, String msg) {
        if ((System.currentTimeMillis() - date.getTime()) > 1000 || TextUtils.isEmpty(lastDateFormatted)) {
            date.setTime(System.currentTimeMillis());
            lastDateFormatted = simpleDateFormat.format(date);
            resetTimePrefix();
        }
        return formatString(level, tag, msg);
    }
}

File: library_zap/src/main/java/com/suyghur/dolin/zap/impl/DateFileFormatter.kt
package com.suyghur.dolin.zap.impl

import android.text.TextUtils
import com.suyghur.dolin.zap.entity.Level
import com.suyghur.dolin.zap.internal.IFormatter
import com.suyghur.dolin.zap.util.LevelUtils
import java.text.SimpleDateFormat
import java.util.*

/**
 * @author #Suyghur.
 * Created on 4/8/21
 */
class DateFileFormatter : IFormatter {

    private val pattern = "yyy-MM-dd HH:mm:ss"
    private var simpleDateFormat: SimpleDateFormat = SimpleDateFormat(pattern, Locale.getDefault())
    private var date: Date
    private var lastDateFormatted = ""
    private var stringBuilder: StringBuilder = StringBuilder()
    private var timeLength = 0

    init {
        Calendar.getInstance()[Calendar.SECOND] = 0
        date = Calendar.getInstance().time
    }

    private fun resetTimePrefix() {
        if (stringBuilder.isNotEmpty()) {
            stringBuilder.delete(0, stringBuilder.length)
        }
        timeLength = stringBuilder.append(lastDateFormatted).append(" ").length
    }

    private fun formatString(level: Level, tag: String, msg: String): String {
        if (stringBuilder.length > timeLength) {
            stringBuilder.delete(timeLength, stringBuilder.length)
        }
        return stringBuilder.append("[${LevelUtils.getShortLevelName(level)}").append(tag).append(": ").append(msg).append("\n").toString()
    }

    override fun format(level: Level, tag: String, msg: String): String {
        if ((System.currentTimeMillis() - date.time) > 1000 || TextUtils.isEmpty(lastDateFormatted)) {
            date.time = System.currentTimeMillis()
            lastDateFormatted = simpleDateFormat.format(date)
            resetTimePrefix()
        }
        return formatString(level, tag, msg)
    }
}